1610520/remove_field.py

```diff
@@ -0,0 +1,214 @@
+fields from the raw crash file in S3 and the document in Elasticsearch.
+
+Usage:
+
+    python 1610520/remove_field.py CRASHIDSFILE FIELD [FIELD...]
+
+"""
+
+import argparse
+import concurrent.futures
+from functools import partial
+import io
+import json
+import logging
+import os
+import sys
+
+from botocore.client import ClientError
+from configman import ConfigurationManager
+from configman.environment import environment
+from elasticsearch.exceptions import ConnectionError
+from elasticsearch_dsl import Search
+from more_itertools import chunked
+
+from socorro.external.boto.connection_context import S3Connection
+from socorro.external.boto.crashstorage import dict_to_str
+from socorro.external.es.connection_context import ConnectionContext
+from socorro.lib.ooid import date_from_ooid
+from socorro.lib.util import retry
+
+
+# Number of seconds until we decide a worker has stalled
+WORKER_TIMEOUT = 15 * 60
+
+# Number of crashids to hand to a worker to process in a single batch
+CHUNK_SIZE = 1000
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_s3_context():
+    """Return an S3ConnectionContext."""
+    cm = ConfigurationManager(
+        S3Connection.get_required_config(), values_source_list=[environment]
+    )
+    config = cm.get_config()
+    return S3Connection(config)
+
+
+def get_es_conn():
+    """Return an Elasticsearch ConnectionContext."""
+    cm = ConfigurationManager(
+        ConnectionContext.get_required_config(), values_source_list=[environment]
+    )
+    config = cm.get_config()
+    return ConnectionContext(config)
+
+
+def crashid_generator(fn):
+    """Lazily yield crash ids."""
+    with open(fn, "r") as fp:
+        for line in fp:
+            line = line.strip()
+            if line.startswith("#"):
+                continue
+            yield json.loads(line)
+
+
+def wait_times_access():
+    """Return generator for wait times between failed load/save attempts."""
+    for i in [5, 5, 5, 5, 5]:
+        yield i
+
+
+@retry(
+    retryable_exceptions=[ClientError],
+    wait_time_generator=wait_times_access,
+    module_logger=logger,
+)
+def fix_data_in_s3(fields, bucket, s3_client, crash_data):
+    """Fix data in raw_crash file in S3."""
+    crashid = crash_data["crashid"]
+
+    path = (
+        "v2/raw_crash/%(entropy)s/%(date)s/%(crashid)s"
+        % {
+            "entropy": crashid[:3],
+            "date": date_from_ooid(crashid).strftime("%Y%m%d"),
+            "crashid": crashid
+        }
+    )
+    resp = s3_client.get_object(Bucket=bucket, Key=path)
+    raw_crash_as_string = resp["Body"].read()
+    data = json.loads(raw_crash_as_string)
+    should_save = False
+    for field in fields:
+        if field in data:
+            del data[field]
+            should_save = True
+
+    if should_save:
+        s3_client.upload_fileobj(
+            Fileobj=io.BytesIO(dict_to_str(data).encode("utf-8")), Bucket=bucket, Key=path
```

Comment:
I learned this week that `utf-8` is the default for `.encode()`. I'm trying to decide if 2020 is the year to use the default, or to keep it explicit.

Reply:
Interesting! I think that changed with Python 3. I remember the default being whatever the encoding in the environment was set to and one could never rely on anything and it sucked.